sympy/functions/elementary/piecewise.py

```diff
@@ -440,7 +440,7 @@ def _handle_irel(self, x, handler):
             # if any condition reduced to True, it needs to go last
             # and there should only be one of them or else the exprs
             # should agree
-            trues = [i for i in range(len(args)) if args[i][1] is S.true]
+            trues = [i for i in range(len(args)) if args[i][1] is True]
```

Comment:
Above args is created with things like `Or` and `And` that will give `S.true` rather than `True`.

What inputs would we need to use to get those cases?

Reply:
Oh, my bad. I assumed we were getting `True` as conditions in `cond`, due to arguments passed to that function in this line : 

https://github.com/sympy/sympy/blob/723630c0296d4d3ef32f29d3bb0c1fd68f0c9999/sympy/functions/elementary/piecewise.py#L422

It seems that the main issue at hand here is that `_handle_irel` is not being able to handle the kind of functions which will return `S.true` in `conds` ( the function given in the issue seems to be one such case ). It would seem putting `S.true` here will be most appropriate as you initially suggested.

In that case I would suggest moving the appended condition (#L349) after the `if-else` block. Something like  : 

`trues = [i for i in range(len(args)) if args[i][1] is S.true]`
`if trues:`
`__assert len({e for e, c in [args[i] for i in trues]}) == 1`
`__args.append(args.pop(trues.pop()))`
`__while trues:`
`____args.pop(trues.pop())`
`args.append((expr, S.true))`
`return Piecewise(*args)`